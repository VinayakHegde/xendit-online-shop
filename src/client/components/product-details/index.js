export * from './product-details-container';
