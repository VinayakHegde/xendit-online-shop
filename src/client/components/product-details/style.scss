.ProductDetails {
  display: grid;
  grid-gap: 20px;
  gap: 20px;
  grid-template-columns: 300px 1fr;
  &__image{
    width: 300px;
  }
  &__info {
    &:first-child {
      display: grid;
    }
    &__description {
      text-align: justify;
    }
  }
  &__button {
    height: 30px;
    padding: 0;

  }
}
