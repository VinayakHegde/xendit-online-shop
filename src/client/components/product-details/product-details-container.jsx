import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link, useParams } from "react-router-dom";
import { getProducts } from "../../store/product-slice";
import { Button } from "../button";
import { ProductDetails } from "./product-details";
import './style.scss';

export const ProductDetailsContainer = () => {
  const { id } = useParams();
  const { product, noSuchProduct } = useSelector(state => {
    const { data } = state.products;

    return {
      product: data.find(p => p.id === id),
      noSuchProduct: data.length > 0
    };
  });
  const dispatch = useDispatch();
  if (!noSuchProduct) {
    return (
      <>
        <div>Product you looking for does not exist!</div>
        <Link to="/" className="button">
          <Button type="button" title="Return Home" />
        </Link>
      </>
    );
  }
  if (!product) {
    dispatch(getProducts());
    // if product not found, then make api cann to get that product
    return <div>Loading...</div>;
  }
  return (
    <ProductDetails product={product} />
  );
}
