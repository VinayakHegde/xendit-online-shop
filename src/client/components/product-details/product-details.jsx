import React from "react";
import { AddToCart } from "../add-to-cart";
import { ProductImage, ProductStock, ProductDescription, ProductName, ProductPrice } from "../product";
import './style.scss';

export const ProductDetails = ({ product = {
  image: '',
  quantity: '',
  price: 0,
  description: '',
} }) => (
  <section className="ProductDetails">
    <div className="ProductDetails__info">
      <ProductImage src={product.image} classPrefix="ProductDetails" />
      <ProductStock quantity={product.quantity} />
      <ProductPrice price={product.price} classPrefix="ProductDetails" />
      <AddToCart product={product} classPrefix="ProductDetails" />
    </div>
    <div className="ProductDetails__info">
      <ProductName name={product.name} classPrefix="ProductDetails" tag="h2" />
      <ProductDescription descr={product.description} classPrefix="ProductDetails" />
    </div>
  </section >
);
