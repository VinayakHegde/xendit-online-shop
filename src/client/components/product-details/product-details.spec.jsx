import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import * as ReactRedux from 'react-redux';
import { fromString } from "uuidv4";
import { ProductDetails } from "./product-details";

afterAll(() => {
  jest.clearAllMocks();
});
const spyOnUseDispatch = (mock, expected) =>
  jest.spyOn(ReactRedux, 'useDispatch').mockImplementation(() => mock.mockResolvedValue(expected));

const getProduct = (props = {}) => {
  return {
    product: {
      id: fromString('Sunglass'),
      name: 'Sunglass',
      price: 9299,
      image: 'sunglasses.png',
      quantity: 10,
      description: 'Sunglasses or sun glasses (informally called shades or sunnies) are a form of protective eyewear designed primarily to prevent bright sunlight and high-energy visible light from damaging or discomforting the eyes. They can sometimes also function as a visual aid, as variously termed spectacles or glasses exist, featuring lenses that are colored, polarized or darkened. In the early 20th century, they were also known as sun cheaters',
      ...props
    }
  }
}
describe.each([
  ["with default stock", getProduct()],
  ["with out of stock", getProduct({ quantity: 0 })],
  ["with low stock", getProduct({ quantity: 2 })]
])('ProductList %s', (a, props) => {
  it('should render without crash', () => {
    jest.spyOn(ReactRedux, 'useSelector').mockImplementation(() => { });
    spyOnUseDispatch(jest.fn(), {});
    const { container } = render(<BrowserRouter><ProductDetails {...props} /></BrowserRouter>)
    expect(container).toMatchSnapshot();
  });
});
