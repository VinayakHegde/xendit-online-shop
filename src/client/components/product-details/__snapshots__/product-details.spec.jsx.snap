// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList with default stock should render without crash 1`] = `
<div>
  <section
    class="ProductDetails"
  >
    <div
      class="ProductDetails__info"
    >
      <img
        alt=""
        class="ProductDetails__image"
        src="/images/sunglasses.png"
      />
      <div
        class="ProductDetails__info__price"
      >
        <span
          class="ProductDetails__label"
        >
          Price
          : 
        </span>
        <span
          class="ProductDetails__price"
        >
          9299
        </span>
      </div>
      <button
        class="Button AddToCart ProductDetails__button"
        title="Add to cart"
        type="button"
      >
        Add to cart
      </button>
    </div>
    <div
      class="ProductDetails__info"
    >
      <h2
        class="ProductDetails__name"
      >
        Sunglass
      </h2>
      <div
        class="ProductDetails__info__description"
      >
        Sunglasses or sun glasses (informally called shades or sunnies) are a form of protective eyewear designed primarily to prevent bright sunlight and high-energy visible light from damaging or discomforting the eyes. They can sometimes also function as a visual aid, as variously termed spectacles or glasses exist, featuring lenses that are colored, polarized or darkened. In the early 20th century, they were also known as sun cheaters
      </div>
    </div>
  </section>
</div>
`;

exports[`ProductList with low stock should render without crash 1`] = `
<div>
  <section
    class="ProductDetails"
  >
    <div
      class="ProductDetails__info"
    >
      <img
        alt=""
        class="ProductDetails__image"
        src="/images/sunglasses.png"
      />
      <div
        class="ProductStock__low"
      >
        <small>
          only 
          2
           left
        </small>
      </div>
      <div
        class="ProductDetails__info__price"
      >
        <span
          class="ProductDetails__label"
        >
          Price
          : 
        </span>
        <span
          class="ProductDetails__price"
        >
          9299
        </span>
      </div>
      <button
        class="Button AddToCart ProductDetails__button"
        title="Add to cart"
        type="button"
      >
        Add to cart
      </button>
    </div>
    <div
      class="ProductDetails__info"
    >
      <h2
        class="ProductDetails__name"
      >
        Sunglass
      </h2>
      <div
        class="ProductDetails__info__description"
      >
        Sunglasses or sun glasses (informally called shades or sunnies) are a form of protective eyewear designed primarily to prevent bright sunlight and high-energy visible light from damaging or discomforting the eyes. They can sometimes also function as a visual aid, as variously termed spectacles or glasses exist, featuring lenses that are colored, polarized or darkened. In the early 20th century, they were also known as sun cheaters
      </div>
    </div>
  </section>
</div>
`;

exports[`ProductList with out of stock should render without crash 1`] = `
<div>
  <section
    class="ProductDetails"
  >
    <div
      class="ProductDetails__info"
    >
      <img
        alt=""
        class="ProductDetails__image"
        src="/images/sunglasses.png"
      />
      <div
        class="ProductStock__outofstock"
      >
        <small>
          out of stock
        </small>
      </div>
      <div
        class="ProductDetails__info__price"
      >
        <span
          class="ProductDetails__label"
        >
          Price
          : 
        </span>
        <span
          class="ProductDetails__price"
        >
          9299
        </span>
      </div>
      <button
        class="Button AddToCart ProductDetails__button"
        disabled=""
        title="Add to cart"
        type="button"
      >
        Add to cart
      </button>
    </div>
    <div
      class="ProductDetails__info"
    >
      <h2
        class="ProductDetails__name"
      >
        Sunglass
      </h2>
      <div
        class="ProductDetails__info__description"
      >
        Sunglasses or sun glasses (informally called shades or sunnies) are a form of protective eyewear designed primarily to prevent bright sunlight and high-energy visible light from damaging or discomforting the eyes. They can sometimes also function as a visual aid, as variously termed spectacles or glasses exist, featuring lenses that are colored, polarized or darkened. In the early 20th century, they were also known as sun cheaters
      </div>
    </div>
  </section>
</div>
`;
